include/exadg/structure/spatial_discretization/operator.cpp

```diff
@@ -662,6 +662,11 @@ Operator<dim, Number>::evaluate_nonlinear_residual(VectorType &       dst,
     body_force_operator.evaluate_add(body_forces, src, time);
     dst -= body_forces;
   }
+
+  // check that constrained degrees of freedom are really zero
+  AssertThrow(check_constrained_values_are_zero(dst),
```

Comment:
Do we really need this in release mode?

Reply:
One could probably think about both options. The reason why I originally decided for AssertThrow is that the costs should be small (note that vmult() of the linear operator touches the constraints before and after matrix-vector product), so I thought these costs are negligible since this is only done in every nonlinear residual evaluation (see description of PR). So I would change to Assert only if you and @kronbichler have concerns in terms of computational costs.

Reply:
The costs are likely small (<1%) in all cases we are using right now, but I would still opt to only use it in debug mode (i.e., use `Assert`):
- The check is a pure verification check to be used during algorithm development and nothing that can should ever go wrong in an application context.
- In case algorithms get extended in the future that would increase the frequency of the evaluation of residual (I think of nonlinear Krylov solvers), the cost could get visible.
- The cost might be low here, but I think the example could indicate to a less experienced programmer that "checks on individual vector elements" in the inner workings of an algorithm are the recommended way also for release builds, which is not true in general.

Note that I am a big believer of the `Assert` category of checks as there are many conditions one would like to make extra checks on during development, but then would be too expensive / superfluous to be checked in a working code with release qualification afterwards. This seems to be a prime example of it, even if the cost is negligible right now.